File: _examples/java-example/src/test/java/io/toolisticon/kotlin/avro/example/java/BankAccountCreatedTest.java
package io.toolisticon.kotlin.avro.example.java;

import io.toolisticon.example.bank.BankAccountCreated;
import io.toolisticon.kotlin.avro.repository.AvroSchemaResolverMap;
import org.javamoney.moneta.Money;
import org.junit.jupiter.api.Test;

import java.util.UUID;

import static io.toolisticon.kotlin.avro.codec.SpecificRecordCodec.specificRecordSingleObjectDecoder;
import static io.toolisticon.kotlin.avro.codec.SpecificRecordCodec.specificRecordSingleObjectEncoder;
import static org.assertj.core.api.Assertions.assertThat;

class BankAccountCreatedTest {

  @Test
  void encodeAndDecodeEventWithMoneyLogicalType() {
    //CustomerId customerId = CustomerId.of("1");
    final BankAccountCreated bankAccountCreated = BankAccountCreated.newBuilder()
      .setAccountId(UUID.randomUUID())
      .setCustomerId("1")
      .setInitialBalance(Money.of(100.123456, "EUR"))
      .build();
    final var resolver = new AvroSchemaResolverMap(BankAccountCreated.getClassSchema());

    final var encoded = specificRecordSingleObjectEncoder().encode(bankAccountCreated);

    final BankAccountCreated decoded = (BankAccountCreated) specificRecordSingleObjectDecoder(resolver).decode(encoded);

    assertThat(decoded.getAccountId()).isEqualTo(bankAccountCreated.getAccountId());
    assertThat(decoded.getCustomerId()).isEqualTo("1");
    assertThat(decoded.getInitialBalance()).isEqualTo(Money.of(100.12, "EUR"));
  }
}

File: _examples/kotlin-example/src/test/kotlin/BankAccountCreatedTest.kt
package io.toolisticon.avro.kotlin.example

import io.toolisticon.avro.kotlin.avroSchemaResolver
import io.toolisticon.avro.kotlin.codec.GenericRecordCodec
import io.toolisticon.avro.kotlin.example.customerid.CustomerId
import io.toolisticon.avro.kotlin.example.money.MoneyLogicalType
import io.toolisticon.avro.kotlin.value.CanonicalName
import io.toolisticon.avro.kotlin.value.Name
import org.assertj.core.api.Assertions.assertThat
import org.javamoney.moneta.Money
import org.junit.jupiter.api.Test
import java.util.*

internal class BankAccountCreatedTest {

  @Test
  fun `show schema`() {
    val schema = KotlinExample.avro.schema(BankAccountCreated::class)

    assertThat(schema.canonicalName).isEqualTo(CanonicalName(BankAccountCreated::class.qualifiedName!!))
    assertThat(schema.fields).hasSize(3)

    val props = schema.getField("initialBalance")?.properties
    assertThat(schema.getField(Name("initialBalance"))?.schema?.logicalTypeName).isEqualTo(MoneyLogicalType.name)
  }

  @Test
  fun `serialize single object with uuid customerId and money`() {
    val amount = Money.of(10, "EUR")
    val accountId = UUID.randomUUID()
    val customerId = CustomerId.random()

    val event = BankAccountCreated(accountId, customerId, amount)
    val resolver = avroSchemaResolver(KotlinExample.avro.schema(BankAccountCreated::class))

    val record = KotlinExample.avro.toRecord(event)

    val json = GenericRecordCodec.encodeJson(record)

    val decodedRecord = GenericRecordCodec.decodeJson(json, KotlinExample.avro.schema(BankAccountCreated::class))

    assertThat(KotlinExample.avro.fromRecord(record, BankAccountCreated::class)).isEqualTo(event)
  }
}
